cmake/Modules/MacroQbtCXXCompilerAndModeCheck.cmake

```diff
@@ -0,0 +1,47 @@
+macro(qbt_cxx_compiler_and_mode_check)
+    # some compilers advertise support for the minimum required version (currently 14),
+    # but it may not be full support
+    # Refer to the C++ mode support matrices on the websites of each compiler
+    set(QBITTORRENT_MINIMUM_CXX_STANDARD "14")
+    set(QBITTORRENT_MINIMUM_GCC_VERSION "5.0")
```

Comment:
Why are these needed if you request C++ standard support? CMake knows what minimal compiler versions are required to fully support the requested standard version. You can use [compile-features]( https://cmake.org/cmake/help/git-master/manual/cmake-compile-features.7.html) if you want a finer control.

Reply:
I thought about using that, but ultimately decided against it. For example, some GCC 4.x advertise C++14 support, but to check if it really has full C++support, I would have to check against the 4 C++14 features that were only added in 5.0. So I thought I might as well just check the version.

Reply:
I think you are wrong. CMake already maintains a table with minimal compiler versions for each standard level.

Reply:
@zeule 
I tested with GCC/G++ 4.8.5 on Ubuntu; without the version checks (i.e. everything inside the `if ("cxx_std_${QBITTORRENT_MINIMUM_CXX_STANDARD}" IN_LIST CMAKE_CXX_COMPILE_FEATURES)`), the configure completes successfully, but the build then fails with `unknown flag -std=c++14`.

The actual failure reason in this instance is because GCC 4.8 would expect something like `-std=c++1x` or something, but it also tells us GCC 4.8 passes CMake's `cxx_std_*` check, without actually having full C++14 support, according to the table at https://gcc.gnu.org/projects/cxx-status.html#cxx14. In other words, even if the `-std=` flag name was corrected to be the legacy one, the build would eventually fail as soon as a C++14 feature only introduced with GCC 5.0 were used.

Reply:
https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD_REQUIRED.html#prop_tgt:CXX_STANDARD_REQUIRED

Reply:
Oh, sorry. You set it below. And you say it does not report an error at compile time?

Reply:
@zeule 
> Oh, sorry. You set it below. And you say it does not report an error at compile time?

No error at _configure_ time (as in CMake's configuration). Eventually an error occurs at compile time, since the required features are actually not all present, even though CMake thinks they are (because the compiler mislead it into thinking that).

Reply:
Ah, they advertise C++14 support for GCC 4.8.5, which has a bare minimum but provides a flag to turn it on.  CMake defines the `CMAKE_CXX14_STANDARD__HAS_FULL_SUPPORT` variable when it presumes the compiler support the standard completely. Maybe you can use that?

Reply:
@FranciscoPombal, maybe you did not know:  
`${CMAKE_CXX${QBITTORRENT_MINIMUM_CXX_STANDARD}_STANDARD__HAS_FULL_SUPPORT}` is how weird it looks in CMake syntax.

Reply:
@zeule
Thanks, your solution worked and greatly simplified the code, but the `IF` condition became `CMAKE_CXX${QBITTORRENT_MINIMUM_CXX_STANDARD}_STANDARD__HAS_FULL_SUPPORT` (notice the lack of the outer-most `${}`), due to the way variable expansion occurs inside `IF` statements.